### PR TITLE
dpo collation/padding

### PR DESCRIPTION
**Pull request by winglian**
# Description

When using DPO/ORPO w Qwen 3.5, the flash linear attention autotunes when there are new sequence lengths. by padding to buckets, we can minimize the excessive memory for autotuning without having to pad everything out to max length.

## Release Notes

* **New Features**
  * Added configurable `pad_to_multiple_of` option to customize batch padding behavior during training
  * Enhanced support for DPO, ORPO, IPO, and SIMPO training modes with improved data collation

* **Improvements**
  * Optimized tensor operations for improved performance during training
  * Extended trainer configuration schema for better control over padding strategies

### CODE REVIEW

File: src/axolotl/core/builders/rl.py [hunk 1]
```diff
@@ -228,9 +228,46 @@ def _build_training_arguments(self, total_num_steps):
...
+            return None
+
+        from axolotl.utils.collators.dpo import AxolotlDPODataCollatorWithPadding
+
+        LOG.info(
+            f"Using AxolotlDPODataCollatorWithPadding with pad_to_multiple_of="
+            f"{pad_to_multiple_of}"
+        )
+        return AxolotlDPODataCollatorWithPadding(
+            pad_token_id=self.tokenizer.pad_token_id,
```
Comment (NanoCode012) [inline src/axolotl/core/builders/rl.py:260]: Would this raise any issue given they migrated to self.processor/self.processing_class?

File: src/axolotl/utils/collators/dpo.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    pad_to_multiple_of: int | None = None
+
+    def __call__(self, features: list[dict[str, Any]]) -> dict[str, Any]:
+        pad_to_mult = self.pad_to_multiple_of
+
+        padded_batch: dict[str, Any] = {}
+        for k in features[0].keys():
+            if k.endswith(
+                ("_input_ids", "_attention_mask", "_labels", "_pixel_values")
```
Comment (NanoCode012) [inline src/axolotl/utils/collators/dpo.py:43]: Is there a list we can find for keys to pad? Do we need to also include other multi-modal keys? mm_token_ids etc or is this enough?

File: src/axolotl/utils/collators/dpo.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            ):
+                if self.is_encoder_decoder:
+                    to_pad = [torch.LongTensor(ex[k]) for ex in features]
```
Comment (NanoCode012) [inline src/axolotl/utils/collators/dpo.py]: Should check dtype

File: src/axolotl/utils/schemas/config.py [hunk 4]
```diff
@@ -663,6 +663,12 @@ class AxolotlInputConfig(
             "description": "Pad inputs so each step uses constant sized buffers. This will reduce memory fragmentation and may prevent OOMs, by re-using memory more efficiently. Defaults to True if `sample_packing` enabled"
         },
     )
+    pad_to_multiple_of: int | None = Field(
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/config.py:666]: Potential conflict: KD has the same config name. Should be safe for now.

Comment (NanoCode012) [reply src/axolotl/utils/schemas/config.py:666]: Additionally, we used to just base this off sequence_len, would we need to refactor those handling? It could break any existing tokenized dataset though.

File: src/axolotl/core/builders/rl.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/core/builders/rl.py:260]: I think this is a follow on fix as the rl builder only seems to support self.tokenizer atm.
